### ISSUE TITLE
Spinwait until pod is ready causing telemetry without Kubernetes properties

### ISSUE THREAD

**Issue by sekiyak** [microsoft/ApplicationInsights-Kubernetes#231]
There is a spinwait implementation initializing the Kubernetes properties for telemetry only after the pods get into ready state (https://github.com/microsoft/ApplicationInsights-Kubernetes/blob/acbcbeb30d85ed2a7a8b2365b77d63aac82ba241/src/ApplicationInsights.Kubernetes/K8sEnvironmentFactory.cs#L160).

What is the reasoning behind this design? Currently, some of our application telemetry is missing the Kubernetes properties due to this causing inconsistencies, as there is a delay during startup until the pods are ready; but there is still telemetry collected by AppInsights auto-collectors.

**Comment by xiaomi7732** [comment]
Hi @sekiyak, thanks for the feedback. Don't you have a concern that when the pod is not ready, the information there might be intermediate and end up pollute the diagnostic information?

**Comment by sekiyak** [comment]
1. Yes, if the k8s properties are available before they are in ready state, it would be ideal if we can enrich the telemetry with them even before the ready state (best-effort if there is any chance they won't be available).
2. This is not a blocking issue for us, but until the pods are marked ready our applications are already live making requests during startup, i.e. health checks, getting identity tokens, loading certificates from KeyVault, etc. and it would be nice to have visibility into such telemetry at the pod level.
3. I'm not familiar with that, if there is indeed such a limitation then I guess there is nothing we can do about it.

The telemetry during startup is useful regardless if it is enriched with k8s properties or not. If we still can enrich them with k8s properties, then that would be ideal.

**Comment by xiaomi7732** [comment]
The current implementation won't block application insights events even when they are not enhanced. If the events aren't there, there's other things going on. A quick way to tell: put application insights SDK on without enable the enhancer.

I'll close this conversation for now due to no actionable item. Feel free to start new issue, referencing this one if related.

@sekiyak thanks for the discussion.
